src/plz/plz.go

```diff
@@ -39,34 +39,44 @@ func Run(targets, preTargets []core.BuildLabel, state *core.BuildState, config *
 	// Start looking for the initial targets to kick the build off
 	go findOriginalTasks(state, preTargets, targets, arch)
 
-	parses, actions, remoteActions := state.TaskQueues()
+	parses, actions := state.TaskQueues()
+
+	localLimiter := make(limiter, config.Please.NumThreads)
+	remoteLimiter := make(limiter, config.NumRemoteExecutors())
 
 	// Start up all the build workers
 	var wg sync.WaitGroup
-	wg.Add(1 + config.Please.NumThreads + config.NumRemoteExecutors())
+	wg.Add(2)
 	go func() {
-		// All parses happen async on separate goroutines so we don't have to worry about them blocking.
-		// They manage concurrency control themselves.
 		for task := range parses {
 			go func(task core.ParseTask) {
-				parse.Parse(0, state, task.Label, task.Dependent, task.ForSubinclude)
+				parse.Parse(state, task.Label, task.Dependent, task.ForSubinclude)
+				state.TaskDone()
+			}(task)
+		}
+		wg.Done()
+	}()
+	go func() {
+		for task := range actions {
+			go func(task core.Task) {
+				if task.Remote {
+					remoteLimiter.Acquire()
+					defer remoteLimiter.Release()
+				} else {
+					localLimiter.Acquire()
+					defer localLimiter.Release()
+				}
```

Comment:
Really excited to see if we can defer this decision into build_step, and build locally, but spill over into remote execution if there's too much backpressure on the local limiter.

Reply:
I think it's possible - definitely more possible than before. The limiter would need to be quite a bit smarter though.